lib/puppet/util/windows/service.rb

```diff
@@ -266,34 +293,41 @@ def exists?(service_name)
     end
     module_function :exists?
 
-    # Start a windows service, assume that the service is already in the stopped state
+    # Start a windows service
     #
     # @param [:string] service_name name of the service to start
     def start(service_name)
-      open_service(service_name, SC_MANAGER_CONNECT, SERVICE_START | SERVICE_QUERY_STATUS) do |service|
-        transition_service_state(service, SERVICE_STOP_PENDING, SERVICE_STOPPED, SERVICE_START_PENDING, SERVICE_RUNNING) do
-          if StartServiceW(service, 0, FFI::Pointer::NULL) == FFI::WIN32_FALSE
-            raise Puppet::Util::Windows::Error.new(_("Failed to start the service"))
-          end
+      Puppet.debug _("Starting the %{service_name} service") % { service_name: service_name }
+
+      valid_initial_states = [
+        SERVICE_STOP_PENDING,
+        SERVICE_STOPPED,
+        SERVICE_START_PENDING
+      ]
+
+      transition_service_state(service_name, valid_initial_states, SERVICE_RUNNING) do |service|
+        if StartServiceW(service, 0, FFI::Pointer::NULL) == FFI::WIN32_FALSE
+          raise Puppet::Util::Windows::Error, _("Failed to start the service")
         end
       end
+
+      Puppet.debug _("Successfully started the %{service_name} service") % { service_name: service_name }
     end
     module_function :start
 
-    # Use ControlService to send a stop signal to a windows service
+    # Stop a windows service
     #
     # @param [:string] service_name name of the service to stop
     def stop(service_name)
-      open_service(service_name, SC_MANAGER_CONNECT, SERVICE_STOP | SERVICE_QUERY_STATUS) do |service|
-        transition_service_state(service, SERVICE_START_PENDING, SERVICE_RUNNING, SERVICE_STOP_PENDING, SERVICE_STOPPED) do
-          FFI::MemoryPointer.new(SERVICE_STATUS.size) do |status_ptr|
-            status = SERVICE_STATUS.new(status_ptr)
-            if ControlService(service, SERVICE_CONTROL_STOP, status) == FFI::WIN32_FALSE
-              raise Puppet::Util::Windows::Error.new(_("Failed to send stop control to service, current state is %{current_state}. Failed with") % { current_state: status[:dwCurrentState].to_s })
-            end
-          end
-        end
+      Puppet.debug _("Stopping the %{service_name} service") % { service_name: service_name }
+
+      valid_initial_states = SERVICE_STATES.keys - [SERVICE_STOPPED]
```

Comment:
Is there a specific reason SERVICE_STOPPED isn't a valid initial state?

(and same question for SERVICE_RUNNING in .start)

Reply:
I was thinking it would make it easier to reason about `transition_service_state` by treating the final state separately from the initial state, i.e. so it's something like:
# If the service is in the final state, no further work needs to be done
# Validate the service's initial state
# If it's already transitioning to the final state, wait then return
# Else, perform the transition and wait

whereas adding the final state as an initial state would make it something like:
# Validate the service's initial state
# If the service is in the final state, no further work needs to be done
# If it's already transitioning to the final state, wait then return
# Else, perform the transition and wait

(i.e. validate should move up, otherwise there would be no point in adding the final state in the set of initial states since we will never reach the validation code).

In other words, an initial state in the first set of semantics is a state that _can_ transition to the final state. Having the final state be an initial state there is kind of weird

Reply:
Alrighty, that makes sense.